### PR TITLE
Perl code

### PR DESCRIPTION
**Pull request by melmothx**
This is a work in progress which is unclear if it will be ever completed... but given that it doesn't touch anything existing and just adds some sample code, you could take it as-is.

### CODE REVIEW

**Comment by ojwb** [comment]
We're certainly happy for partial work towards translating the example code.

However your changes break the build, which is problematic - try running:

```
make clean && make html LANGUAGE=perl
```

In the process of building the documentation this runs the new example code and verifies that the output matches the output the manual claims is produced, and also extracts snippets from the example code to insert into the generated documentation based on special comments in the the example code.  These currently fail with your changes in several cases.  For example:

* the output format of `code/perl/search_facets.pl` adds the maker name which isn't present in the other versions

* `code/perl/search_ranges.pl` dies with:

```
Xapian::Search::NumberRangeProcessor is not bound in perl! at code/perl/search_ranges.pl line 33.
```

* `code/perl/index_filters.pl` is missing the `### End of new indexing code.` special comment which tells the Sphinx `xapianexample` macro where to stop extracting example code to insert into the documentation.

**Comment by melmothx** [comment]
It should be good to go now.

```
rm -rf _build/*
rm -rf db statesdb
rm -rf code/c++/built
rm -f code/python/*.pyc code/python3/*.pyc
sphinx-build -b html -q -d _build/doctrees   -tperl . _build/html
/home/melmoth/amw/xapian/xapian-docsprint/concepts/search/search_limitations.rst:46: WARNING: unknown document: /advanced/remote
/home/melmoth/amw/xapian/xapian-docsprint/index.rst:14: WARNING: toctree contains reference to document u'language_specific' that doesn't have a title: no link will be generated
*** No version of example postingsource in language perl - patches welcome!
*** No version of example search_ranges in language perl - patches welcome!
*** No version of example index_ranges2 in language perl - patches welcome!
*** No version of example search_ranges2 in language perl - patches welcome!
*** No version of example search_sorting in language perl - patches welcome!
*** No version of example search_sorting2 in language perl - patches welcome!
*** No version of example index_values_with_geo in language perl - patches welcome!
*** No version of example search_sorting3 in language perl - patches welcome!
*** No version of example search_synonyms in language perl - patches welcome!

Build finished. The HTML pages are in _build/html.
```

**Comment by melmothx** [comment]
I think you can expect from me to complete the translation in a reasonable time, so please don't merge and let it open. Of course the ranges one can't be done because currently there is no perl binding for that. I'd also add some language specific note about encoding/decoding.

That's, at least, the grand plan.

**Comment by melmothx** [comment]
I think that's it for now:

```
*** No version of example postingsource in language perl - patches welcome!
*** No version of example search_ranges in language perl - patches welcome!
*** No version of example search_ranges2 in language perl - patches welcome!
*** No version of example index_values_with_geo in language perl - patches welcome!
*** No version of example search_sorting3 in language perl - patches welcome!
```

These are scripts which can't currently work with Search::Xapian because of subclassing or because the class is missing.

File: code/perl/index_filters.pl [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        $term_generator->set_document($doc);
+        # Index each field with a suitable prefix.
+        $term_generator->index_text($rec->{TITLE}, 1, 'S');
+        $term_generator->index_text($rec->{DESCRIPTION}, 1, 'XD');
+
+        ### Start of new indexing code.
+        # Index the MATERIALS field, splitting on semicolons.
+        foreach my $material (split(/;/, $rec->{MATERIALS})) {
+            $material =~ s/\A\s*//;
+            $material =~ s/\z\s*//;
```
Comment (ojwb) [inline code/perl/index_filters.pl]: This removes whitespace **after** the end of the string - I think you meant `s/\s*\z//`.

File: code/perl/search_filters.pl [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    # And parse the query
+    my $query = $queryparser->parse_query($query_string);
+
+    # there is no pod for Search::Xapian::Query, but works anyway. Operator + list.
+
+    if (@$materials) {
+        my $material_query = Search::Xapian::Query->new(+OP_OR,
+                                                        map { Search::Xapian::Query->new('XM' . lc($_)) }
+                                                        @$materials);
+        $query = Search::Xapian::Query->new(+OP_FILTER, $query, $material_query);
```
Comment (ojwb) [inline code/perl/search_filters.pl]: `+` shouldn't be needed before the constants here.

File: code/perl/search_filters2.pl [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+use JSON::MaybeXS;
+use Search::Xapian ':all';
+use FindBin qw($Bin);
+use lib $Bin;
+use Support;
+use Data::Dumper;
+
+
+my ($db_path, $query_string) = @ARGV;
+die "Usage: $0 DB_PATH QUERYTERM..." unless $db_path && $query_string;
```
Comment (ojwb) [inline code/perl/search_filters2.pl]: This ignores any arguments after the second - the code in `search_sorting.py` looks better.

**Comment by ojwb** [comment]
Also this is a substantial contribution so you should probably update `LICENSE.rst` with copyright details.

**Comment by melmothx** [comment]
Please let me know if you want me to write some notes about perl specific encoding/decoding in the language specific section.

**Comment by ojwb** [comment]
> Please let me know if you want me to write some notes about perl specific encoding/decoding in the language specific section.

That sounds useful (in case you've not found it yet, adding some content `language_specific/perl/index.rst` should work).

**Comment by ojwb** [comment]
I've also pushed a changed to `overview.rst` to add Perl to the list of languages there's example code for.

**Pull request merged**